### ISSUE TITLE
[ENH] Prediction intervals for `EnsembleForecaster`

### ISSUE THREAD

**Issue by topher-lo** [alan-turing-institute/sktime#3438]
**Is your feature request related to a problem? Please describe.**
- `EnsembleForecaster` does not have`_predict_interval capabilities

**Describe the solution you'd like**
- Add a check that all underlying forecasters have  `predict_var` or `predict_quantile` capabilities
- To actually implement `_predict_interva`. Two possible solutions (inspired by discussion in Hyndman blog https://robjhyndman.com/hyndsight/forecast-combinations/#forecasthybrid):
   - (Easy way) Take the widest possible intervals that includes all the intervals produced by the individual models. This is the method that the R library [ForecastHybrid](https://robjhyndman.com/hyndsight/forecast-combinations/#forecasthybrid) uses. However, the prediction interval will likely be too conversative (i.e. wide). You only need one bad model and the interval coverage will seem terrible.
   - (Hard way) "Estimate the covariances between the different forecast errors, and then find the resulting variance expression for the linear combination of methods." -- this Rob Hyndman's suggestion (see ForecastHybrid link above). Further research is needed to find the appropriate algorithm for this.

**Describe alternatives you've considered**
See above.

**Additional context**

**Comment by fkiraly** [comment]
I would suggest a `"method"` parameter which allows the user to choose?

3rd option, medium: just apply the aggregation function to the predicted variances, intervals, etc.

4th option: allow the user to specify an aggregation function for the prediction intervals etc (also leaving it up to the user to ensure they are consistent with each other).

**Comment by fkiraly** [comment]
> "Estimate the covariances between the different forecast errors, and then find the resulting variance expression for the linear combination of methods." -- this Rob Hyndman's suggestion (see ForecastHybrid link above). Further research is needed to find the appropriate algorithm for this.

This is for `predict_var` only, and uses Bienaymé's formula with plug-in estimates, I suppose.

**Comment by topher-lo** [comment]
> I would suggest a `"method"` parameter which allows the user to choose?
> 
> 3rd option, medium: just apply the aggregation function to the predicted variances, intervals, etc.
> 
> Should be fine for homoskedastic normal distributions, but might produce unexpected results for very skew or heavy-tailed distributions.
> 
> 4th option: allow the user to specify an aggregation function for the prediction intervals etc (also leaving it up to the user to ensure they are consistent with each other).

A few thoughts:
- An opinated and limited number of methods would be easiest. It also follows the current API which takes a string alias ("mean", "median", "min", "max") for `aggfunc`.
- IMO a general aggregation function is non-trivial (and could be a smell code) for one key reason: unlike point forecasts, ensembling forecast intervals requires either:
   1. The underyling forecaster's "metadata" (e.g. the name of the forecaster)
   2. Or estimates of the **pairwise covariance of forecast residuals** across all ensembled forecasters
- To cover both cases, the input signature for a user provided Callable "aggfunc_interval" will ikely too broad with lots of optional / unused combinations of inputs.
- I would like to stress the point about the "covariances". Unlike #3437, we really should not assume uncorrelated forecast residuals for `EnsembleForecaster`. This means that any linear combination of the variances (without covariance) will give very biased estimates.

My suggestion: we should be opinionated and just support a few "methods" (see next comment for breakdown) for aggregating prediction intervals / variances.

**Comment by topher-lo** [comment]
There seems to be two families of "methods" for aggregating prediction intervals:
1. Simple rule based method e.g. just select the wide possible interval, just select interval from one of the forecasters (Hyndman uses that rule in this [blog post](https://robjhyndman.com/hyndsight/benchmark-combination/).
2. Bienaymé's formula with "plug-in estimates" for variance and pairwise covariances.

For 1. we can have methods ("widest", "{some forecaster key in forecasters: List[Tuple[str, BaseForecaster]]}").
For 2. the algorithm is as follows:
  1. Predict variance for each forecaster in ensemble
  2. Compute forecast residuals (I suggest we take the in-sample residuals) for each forecaster
  3. Compute pairwise covariance between forecast residuals for each forecaster
  4. Apply Bienaymé's formula and assume normality of ensembled residuals

**Comment by sanjayk0508** [comment]
@fkiraly , I'd like to contribute to this enhancement, as it aligns with the `time-to-event modeling, probabilistic supervised learning & probabilistic forecasting, ` project idea, which I'm particularly drawn to.

Here's my opinion:
The 4th option - **Bienayme formula** based approach appears promising. Considering it is used to estimate the variance of a linear combination of random variables, allows to account for the covariance between the forecast errors of different models, which is crucial, especially in scenarios where forecasts exhibit high correlation.

The discussion took place more than a year ago. So as a suggestion, I'm curious if there are different variations or extensions of the Bienayme formula based approach that might be worth considering for our specific use case at this stage.

**Comment by fkiraly** [comment]
@sanjayk0508, yes, this is still relevant - perhaps even more due to the current work on integration with `tsbootstrap` and the related bootstrap error estimates. FYI @benHeid, @astrogilda.

Re approaches, it would be nice to have both as a possible strategy implemented, for probabilistic prediction. There are actually two, the list of four describes one out of two options overall.

**Comment by fkiraly** [comment]
Btw, @sanjayk0508, very much related is this recent PR: https://github.com/sktime/sktime/pull/6052

If you are interested in this, you could look into `predict_proba` - I currently cannot figure out what goes wrong there for hierarchical data, and I also cannot reproduce the failure locally. Perhaps you have an idea?

**Comment by sanjayk0508** [comment]
I see! Just for clarification, when you mention both strategies, are you referring to the `Bienayme formula` and the `simple weighted empirical` approach? Also, could you provide guidance on how I can address/work on this issue? 

> Re approaches, it would be nice to have both as a possible strategy implemented, for probabilistic prediction. There are actually two, the list of four describes one out of two options overall.
